File: src/main/java/com/murtukov/css_to_jss/CssConverter.java
import com.google.common.primitives.Chars;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.stream.Collectors;

public class CssConverter {
    private final StringBuilder buffer = new StringBuilder();
    private final StringBuilder result = new StringBuilder();
    private final char[] stoppers      = new char[]{'{', ':', ';', '}'};
    private final String[] rules       = new String[]{"margin", "padding"};
    private String currentRule;

    public String parse(String css) {
        var scope = 0;

        css = removeCommentBlocks(normalizeWhitespaces(css));

        for (final char c : css.toCharArray()) {
            if ('(' == c) {
                scope++;
            } else if (')' == c) {
                scope--;
            }

            if (Chars.contains(stoppers, c) && scope == 0) {
                result.append(convertToken(buffer.toString(), c));
                buffer.setLength(0);
            } else {
                buffer.append(c);
            }
        }

        return result.toString();
    }

    private String convertToken(String token, char stopper) {
        token = token.trim();

        switch (stopper) {
            case '{': // selector
                return "{\n";
            case ':': // rule name
                return "    " + convertRuleName(token) + ":";
            case ';': // rule value
                return convertRuleValue(token);
            case '}': // end of block
                return "}";
        }

        return "";
    }

    private String convertRuleName(String token) {
        if (token.charAt(0) == '-') {
            return String.format("'%s'", token);
        }

        return this.currentRule = toCamelCase(token);
    }

    private String convertRuleValue(String token) {
        return " " + new CssValue(token.trim()) + ",\n";
    }

    /**
     * Converts a kebab-cased string into a camelCased one.
     */
    private String toCamelCase(String input) {
        var temp = Arrays.stream(input.split("-"))
                .map(s -> Character.toUpperCase(s.charAt(0)) + s.substring(1).toLowerCase())
                .collect(Collectors.joining());

        return Character.toLowerCase(temp.charAt(0)) + temp.substring(1);
    }

    private String removeCommentBlocks(String input) {
        while (true) {
            var start = input.indexOf("/*");
            var end = input.indexOf("*/");

            if (-1 != start && -1 != end) {
                input = input.substring(0, start) + input.substring(end + 2);
            } else {
                break;
            }
        }

        return input;
    }

    private String normalizeWhitespaces(String input) {
        return input.trim().replaceAll("\\s+", " ");
    }

    /**
     * Wraps string into single quotes if it's not numeric.
     */
    private String maybeWrapIntoQuotes(String input) {
        if (isNumeric(input)) {
            return input;
        }

        return "'" + input + "'";
    }

    private String removeSuffix(String input) {
        if (input.endsWith("px")) {
            return input.substring(0, input.length() - 2);
        }

        return input;
    }

    private boolean isNumeric(String str) {
        try {
            Double.parseDouble(str);
            return true;
        } catch(NumberFormatException e){
            return false;
        }
    }

    private String trimQuotes(String input) {
        return input.replaceAll("^'|'$", "")
                    .replaceAll("^\"|\"$", "");
    }

    private ArrayList<String> split(String input, char separator) {
        var buffer = new StringBuilder();
        var parts  = new ArrayList<String>();
        var parenthesis  = 0;
        var singleQuote = false;
        var doubleQuote = false;

        for (final char c : input.toCharArray()) {
            if ('(' == c) {
                parenthesis++;
            } else if (')' == c) {
                parenthesis--;
            }

            if ('\'' == c) {
                singleQuote = !singleQuote;
            }

            if ('"' == c) {
                doubleQuote = !doubleQuote;
            }

            if (separator == c && 0 == parenthesis && !singleQuote && !doubleQuote) {
                parts.add(buffer.toString().trim().replace('\n', ' '));
                buffer.setLength(0);
            } else {
                buffer.append(c);
            }
        }

        if (buffer.length() > 0) {
            parts.add(buffer.toString().trim().replace('\n', ' '));
            buffer.setLength(0);
        }

        return parts;
    }

    private class CssValue {
        private final ArrayList<String> parts;

        CssValue(String input) {
            parts = split(input, ',');
        }

        @Override
        public String toString() {
            var result = new ArrayList<String>();

            for (String part : parts) {
                part = normalizeWhitespaces(part);
                var partResult = new ArrayList<String>();
                var values = split(part, ' ');

                if (currentRule.equals("fontFamily")) {
                    partResult.add(processValue(part));
                } else {
                    for (String value : values) {
                        partResult.add(processValue(value));
                    }
                }

                if (partResult.size() > 1) {
                    var lastItem = values.get(partResult.size() - 1);

                    if (lastItem.equals("!important") && Arrays.asList(rules).contains(currentRule)) {
                        partResult.remove(partResult.size() - 1);
                        result.add("[" + join(partResult, true, false) + ", '!important']");
                    } else {
                        result.add(join(partResult, true, false));
                    }
                } else {
                    result.add(join(partResult, false, false));
                }
            }

            if (result.size() > 1) {
                if (containsArrays(result)) {
                    return join(result, true, true);
                }
                return join(result, true, false);
            }

            return join(result, false, false);
        }

        private String processValue(String input) {
            return maybeWrapIntoQuotes(trimQuotes(removeSuffix(input)).replace("'", "\\'"));
        }

        private String join(ArrayList<String> input, boolean wrap, boolean multiline) {
            if (wrap) {
                if (multiline) {
                    return "[\n" + String.join(",\n", input) + "\n]";
                }

                return "[" + String.join(", ", input) + "]";
            }

            return String.join(", ", input);
        }

        private boolean containsArrays(ArrayList<String> input) {
            for (final var item : input) {
                if (item.charAt(0) == '[') {
                    return true;
                }
            }

            return false;
        }
    }
}

File: src/main/kotlin/org/murtukov/css_to_jss/CssConverter.kt
package org.murtukov.css_to_jss

import org.murtukov.css_to_jss.util.*
import java.util.*
import kotlin.collections.HashMap
import kotlin.properties.Delegates

class CssConverter {
    private val rules   = arrayOf("margin", "padding")
    private var lines   = HashMap<CssProperty, CssValue>()

    private lateinit var lastProperty: CssProperty

    fun convert(css: String): String {
        val normalizedCss    = css.normalizeWhitespaces().removeCommentBlocks()
        val stoppers         = "{:;}"
        var expectedStoppers = "{:"
        var parenthesisScope = 0
        var buffer           = ""

        parse@for (c in normalizedCss) {
            when (c) {
                '(' -> parenthesisScope++
                ')' -> parenthesisScope--
            }

            if (stoppers.contains(c) && 0 == parenthesisScope) {
                if (!expectedStoppers.contains(c)) {
                    return css // return unchanged
                }

                when (c) {
                    '{' -> expectedStoppers = ":"
                    ':' -> {
                        lastProperty = CssProperty(buffer)
                        expectedStoppers = ";";
                    }
                    ';' -> {
                        lines[lastProperty] = CssValue(buffer, lastProperty)
                        expectedStoppers = ":}"
                    }
                    '}' -> break@parse
                }

                buffer = ""
            } else {
                buffer += c
            }
        }

        // Check, if selection was partial


        return generate()
    }

    private fun generate(): String {
        val stringifiedRules = lines
            .map { "${it.key}: ${it.value}," }
            .joinToString("\n")
            .prependIndent()

        return "{\n$stringifiedRules\n}"
    }

    private fun maybeWrapIntoQuotes(input: String): String {
        return if (input.isNumeric()) input else "'$input'";
    }

    private fun String.split(separator: Char): ArrayList<String> {
        var buffer = ""
        val parts = ArrayList<String>()
        var parenthesis = 0
        var singleQuote = false
        var doubleQuote = false

        for (c in this) {
            when (c) {
                '('  -> parenthesis++
                ')'  -> parenthesis--
                '\'' -> singleQuote = !singleQuote
                '"'  -> doubleQuote = !doubleQuote
            }

            if (separator == c && 0 == parenthesis && !singleQuote && !doubleQuote) {
                parts.add(buffer.trim().replace('\n', ' '))
                buffer = ""
            } else {
                buffer += c
            }
        }

        if (buffer.isNotBlank()) {
            parts.add(buffer.trim().replace('\n', ' '))
        }

        return parts
    }

    private inner class CssProperty(name: String) {
        val cssName = name
        val jssName = when {
            cssName.startsWith('-') -> "'$cssName'"
            else -> cssName.toCamelCase()
        }

        override fun toString() = jssName
    }

    private enum class ValueType {
        SCALAR,                 // 200px
        SPACE_SEPARATED,        // 0 0 20px 30px
        COMMA_SEPARATED,        // "Helvetica Neue", Arial, sams-serif
        COMMA_SPACE_SEPARATED,  // rgba(50, 50, 93, 0.25) 0px 50px 100px -20px, rgba(0, 0, 0, 0.3) 0px 30px 60px -30px
        MIXED                   // italic bold .8em/1.2 Helvetica Neue, Arial, sans-serif
    }

    private inner class CssValue(rawValue: String, private val property: CssProperty) {
        private val parts = rawValue.split(',')
        private var isImportant = false
        private var defaultUnit = "px"
        private val valueType: ValueType = ValueType.SCALAR

        init {
            if (rawValue.contains("!important")) {
                rawValue.removeSuffix("!important")
                isImportant = true
            }
        }

        /**
         * 1. check if !important is set
         * 2. Determine the value type
         */

        override fun toString(): String {
            val result = ArrayList<String>()

            for (part in parts) {
                val normalized = part.normalizeWhitespaces()
                val partResult = ArrayList<String>()
                val values     = normalized.split(' ')

                if ("fontFamily" == property.jssName) {
                    partResult.add(processValue(normalized))
                } else {
                    for (value in values) {
                        partResult.add(processValue(value))
                    }
                }

                if (partResult.size > 1) {
                    if (values.last() == "!important" && rules.contains(property.jssName)) {
                        partResult.removeAt(partResult.size - 1)
                        result.add("[${partResult.join(true)}, '!important']")
                    } else {
                        result.add(partResult.join(true))
                    }
                } else {
                    result.add(partResult.join(false))
                }
            }
            return if (result.size > 1) {
                if (containsArrays(result)) {
                    result.joinToString(",\n", "[\n", "\n]")
                } else result.joinToString(", ", "[", "]")
            } else result.joinToString(", ")
        }

        private fun determineType() {

        }

        private fun processValue(input: String): String {
            return maybeWrapIntoQuotes(
                input.removeSuffix("px")
                    .trimQuotes()
                    .replace("'", "\\'")
            )
        }

        private fun ArrayList<String>.join(wrap: Boolean = false, multiline: Boolean = false): String {
            return when {
                wrap -> {
                    if (multiline)
                        this.joinToString(",\n", "[\n", "\n]")
                    else
                        this.joinToString(", ", "[", "]")
                }
                else -> this.joinToString(", ")
            }
        }

        private fun containsArrays(input: ArrayList<String>): Boolean {
            for (item in input) {
                if (item.startsWith('[')) {
                    return true
                }
            }
            return false
        }
    }
}